File: python/goml/parser.py
"""Recursive descent parser for GOML."""

from __future__ import annotations

from typing import Any, List, Optional

from .ast import (
    ArrayNode,
    ArrayOfObjectsNode,
    BooleanNode,
    GOMLNode,
    KeyValueNode,
    NullNode,
    NumberNode,
    ObjectNode,
    ReferenceNode,
    RootNode,
    StringNode,
)
from .errors import ParseError
from .lexer import Token, TokenType, tokenize


class Parser:
    """Recursive descent parser that produces Python dicts/lists from GOML tokens."""

    def __init__(self, source: str):
        self.tokens = tokenize(source)
        self.pos = 0

    def peek(self, offset: int = 0) -> Token:
        idx = self.pos + offset
        if idx < len(self.tokens):
            return self.tokens[idx]
        return self.tokens[-1]  # EOF

    def advance(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expect(self, tok_type: TokenType) -> Token:
        tok = self.peek()
        if tok.type != tok_type:
            raise ParseError(
                f"Expected {tok_type.name}, got {tok.type.name} ({tok.value!r})",
                tok.line,
                tok.col,
            )
        return self.advance()

    def skip_newlines(self):
        while self.peek().type == TokenType.NEWLINE:
            self.advance()

    def parse_root(self) -> RootNode:
        root = RootNode()
        self.skip_newlines()
        while self.peek().type != TokenType.EOF:
            stmt = self.parse_statement()
            if stmt is not None:
                root.statements.append(stmt)
            self.skip_newlines()
        return root

    def parse_statement(self) -> Optional[GOMLNode]:
        tok = self.peek()

        if tok.type == TokenType.IDENT:
            return self.parse_ident_statement()
        elif tok.type == TokenType.DOLLAR:
            return self.parse_ident_statement()
        else:
            # skip unexpected tokens
            self.advance()
            return None

    def parse_ident_statement(self) -> GOMLNode:
        """Parse key = value, key { ... }, or key [ { ... } ]."""
        if self.peek().type == TokenType.DOLLAR:
            ref_tok = self.advance()
            # Build dotted path: $a.b.c
            path_parts: List[str] = []
            tok = self.peek()
            if tok.type == TokenType.IDENT:
                path_parts.append(tok.value)
                self.advance()
            while self.peek().type == TokenType.DOT:
                self.advance()  # consume dot
                name_tok = self.expect(TokenType.IDENT)
                path_parts.append(name_tok.value)
            ref_path = ".".join(path_parts)
            # Could be $ref = value or $ref alone
            return ReferenceNode(path=ref_path, line=ref_tok.line, col=ref_tok.col)

        key_tok = self.advance()  # IDENT
        key = key_tok.value

        self.skip_newlines()

        next_tok = self.peek()

        if next_tok.type == TokenType.EQUALS:
            self.advance()  # consume =
            self.skip_newlines()
            value = self.parse_value()
            return KeyValueNode(key=key, value=value, line=key_tok.line, col=key_tok.col)

        if next_tok.type == TokenType.LBRACE:
            self.advance()  # consume {
            obj = self.parse_object_body()
            self.expect(TokenType.RBRACE)
            return KeyValueNode(key=key, value=obj, line=key_tok.line, col=key_tok.col)

        if next_tok.type == TokenType.LBRACKET:
            self.advance()  # consume [
            self.skip_newlines()
            if self.peek().type == TokenType.LBRACE:
                # Array of objects: items [ { ... } { ... } ]
                arr = self.parse_array_of_objects()
                self.expect(TokenType.RBRACKET)
                return KeyValueNode(key=key, value=arr, line=key_tok.line, col=key_tok.col)
            else:
                # Normal array
                arr = self.parse_array_items()
                self.expect(TokenType.RBRACKET)
                return KeyValueNode(key=key, value=arr, line=key_tok.line, col=key_tok.col)

        raise ParseError(
            f"Expected '=', '{{', or '[' after key '{key}', got {next_tok.type.name}",
            next_tok.line,
            next_tok.col,
        )

    def parse_value(self) -> GOMLNode:
        tok = self.peek()
        if tok.type == TokenType.STRING:
            self.advance()
            return StringNode(value=tok.value, line=tok.line, col=tok.col)
        if tok.type == TokenType.NUMBER:
            self.advance()
            return NumberNode(value=tok.value, line=tok.line, col=tok.col)
        if tok.type == TokenType.BOOLEAN:
            self.advance()
            return BooleanNode(value=tok.value, line=tok.line, col=tok.col)
        if tok.type == TokenType.NULL:
            self.advance()
            return NullNode(line=tok.line, col=tok.col)
        if tok.type == TokenType.IDENT:
            self.advance()
            return StringNode(value=tok.value, line=tok.line, col=tok.col)
        if tok.type == TokenType.LBRACKET:
            self.advance()
            items = self.parse_array_items()
            self.expect(TokenType.RBRACKET)
            return ArrayNode(items=items, line=tok.line, col=tok.col)
        if tok.type == TokenType.LBRACE:
            self.advance()
            obj = self.parse_object_body()
            self.expect(TokenType.RBRACE)
            return obj
        if tok.type == TokenType.DOLLAR:
            return self.parse_ref_value()
        raise ParseError(
            f"Unexpected token in value position: {tok.type.name} ({tok.value!r})",
            tok.line,
            tok.col,
        )

    def parse_ref_value(self) -> ReferenceNode:
        ref_tok = self.advance()  # consume $
        path_parts: List[str] = []
        tok = self.peek()
        if tok.type == TokenType.IDENT:
            path_parts.append(tok.value)
            self.advance()
        while self.peek().type == TokenType.DOT:
            self.advance()
            name_tok = self.expect(TokenType.IDENT)
            path_parts.append(name_tok.value)
        return ReferenceNode(path=".".join(path_parts), line=ref_tok.line, col=ref_tok.col)

    def parse_object_body(self) -> ObjectNode:
        obj = ObjectNode()
        self.skip_newlines()
        while self.peek().type not in (TokenType.RBRACE, TokenType.EOF):
            stmt = self.parse_statement()
            if stmt is not None:
                obj.entries.append(stmt)
            self.skip_newlines()
        return obj

    def parse_array_items(self) -> List[GOMLNode]:
        items: List[GOMLNode] = []
        self.skip_newlines()
        while self.peek().type not in (TokenType.RBRACKET, TokenType.EOF):
            items.append(self.parse_value())
            self.skip_newlines()
            if self.peek().type == TokenType.COMMA:
                self.advance()  # consume comma
                self.skip_newlines()
        return items

    def parse_array_of_objects(self) -> ArrayOfObjectsNode:
        arr = ArrayOfObjectsNode()
        self.skip_newlines()
        while self.peek().type != TokenType.RBRACKET:
            if self.peek().type == TokenType.LBRACE:
                self.advance()  # consume {
                obj = self.parse_object_body()
                self.expect(TokenType.RBRACE)
                arr.items.append(obj)
            else:
                break
            self.skip_newlines()
        return arr


def _node_to_python(node: GOMLNode) -> Any:
    """Convert an AST node to a native Python value."""
    if isinstance(node, StringNode):
        return node.value
    if isinstance(node, NumberNode):
        return node.value
    if isinstance(node, BooleanNode):
        return node.value
    if isinstance(node, NullNode):
        return None
    if isinstance(node, ReferenceNode):
        return f"__ref__:{node.path}"
    if isinstance(node, ArrayNode):
        return [_node_to_python(item) for item in node.items]
    if isinstance(node, ObjectNode):
        result = {}
        for entry in node.entries:
            if isinstance(entry, KeyValueNode):
                result[entry.key] = _node_to_python(entry.value)
        return result
    if isinstance(node, ArrayOfObjectsNode):
        return [_node_to_python(obj) for obj in node.items]
    if isinstance(node, RootNode):
        # If single top-level key-value, return dict directly
        # If multiple, return dict with all keys
        result = {}
        for stmt in node.statements:
            if isinstance(stmt, KeyValueNode):
                result[stmt.key] = _node_to_python(stmt.value)
        return result
    return None


def loads(source: str) -> dict:
    """Parse GOML string and return a Python dict."""
    parser = Parser(source)
    tree = parser.parse_root()
    result = _node_to_python(tree)
    if not isinstance(result, dict):
        return result or {}
    return result


def loads_all(source: str) -> List[dict]:
    """Parse GOML source with multiple root-level objects and return a list of dicts."""
    parser = Parser(source)
    tree = parser.parse_root()
    results = []
    for stmt in tree.statements:
        if isinstance(stmt, KeyValueNode):
            results.append({stmt.key: _node_to_python(stmt.value)})
    return results

File: js/src/parser.ts
import { Tokenizer, Token, TokenType } from './lexer';
import { ParseError } from './errors';
import {
  Document,
  GomlNode,
  GomlValue,
  KeyValue,
  BlockObject,
  BlockArray,
  CommentNode,
  StringValue,
  NumberValue,
  BooleanValue,
  NullValue,
  ArrayValue,
  InlineObject,
  ReferenceValue,
} from './ast';

export class Parser {
  private tokens: Token[] = [];
  private pos: number = 0;

  parse(input: string): Document {
    const tokenizer = new Tokenizer(input);
    this.tokens = tokenizer.tokenize();
    this.pos = 0;

    const entries: GomlNode[] = [];
    while (!this.isAtEnd()) {
      this.skipNewlinesAndComments(entries);
      if (this.isAtEnd()) break;

      const token = this.current();
      if (token.type === TokenType.Newline || token.type === TokenType.Comment) {
        this.advance();
        continue;
      }

      if (token.type === TokenType.Ident) {
        entries.push(this.parseStatement());
      } else if (token.type === TokenType.RBrace || token.type === TokenType.RBracket) {
        break;
      } else {
        throw new ParseError(
          `Unexpected token '${token.literal}' at ${token.line}:${token.col}`,
          token.line,
          token.col
        );
      }

      this.skipNewlinesAndComments(entries);
    }

    return { type: 'Document', entries };
  }

  private parseStatement(): GomlNode {
    const nameToken = this.expect(TokenType.Ident);
    const name = nameToken.literal;

    this.skipNewlinesAndComments();

    if (this.check(TokenType.LBrace)) {
      return this.parseBlockObject(nameToken);
    }

    if (this.check(TokenType.LBracket)) {
      return this.parseBlockArray(nameToken);
    }

    this.expect(TokenType.Equals);
    this.skipNewlinesAndComments();
    const value = this.parseValue();

    return {
      type: 'KeyValue',
      key: name,
      value,
      line: nameToken.line,
      col: nameToken.col,
    };
  }

  private parseBlockObject(nameToken: Token): BlockObject {
    this.expect(TokenType.LBrace);
    this.skipNewlinesAndComments();

    const entries: GomlNode[] = [];
    while (!this.check(TokenType.RBrace) && !this.isAtEnd()) {
      if (this.check(TokenType.Newline)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.Comment)) {
        entries.push(this.parseComment());
        continue;
      }

      if (this.check(TokenType.RBrace)) break;

      const token = this.current();
      if (token.type === TokenType.Ident) {
        entries.push(this.parseStatement());
        this.skipNewlinesAndComments();
      } else {
        throw new ParseError(
          `Unexpected token '${token.literal}' at ${token.line}:${token.col}`,
          token.line,
          token.col
        );
      }

      this.skipNewlinesAndComments();
    }

    this.expect(TokenType.RBrace);

    return {
      type: 'Object',
      name: nameToken.literal,
      entries,
      line: nameToken.line,
      col: nameToken.col,
    };
  }

  private parseBlockArray(nameToken: Token): BlockArray {
    this.expect(TokenType.LBracket);
    this.skipNewlinesAndComments();

    const items: GomlValue[] = [];
    while (!this.check(TokenType.RBracket) && !this.isAtEnd()) {
      if (this.check(TokenType.Newline)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.Comment)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.RBracket)) break;

      const token = this.current();
      if (token.type === TokenType.LBrace) {
        const obj = this.parseInlineObject();
        items.push(obj);
      } else {
        throw new ParseError(
          `Expected object in array at ${token.line}:${token.col}`,
          token.line,
          token.col
        );
      }

      this.skipComma();
      this.skipNewlinesAndComments();
    }

    this.expect(TokenType.RBracket);

    return {
      type: 'Array',
      name: nameToken.literal,
      items,
      line: nameToken.line,
      col: nameToken.col,
    };
  }

  private parseInlineObject(): InlineObject {
    this.expect(TokenType.LBrace);
    this.skipNewlinesAndComments();

    const entries: KeyValue[] = [];
    while (!this.check(TokenType.RBrace) && !this.isAtEnd()) {
      if (this.check(TokenType.Newline)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.Comment)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.RBrace)) break;

      const token = this.current();
      if (token.type === TokenType.Ident) {
        const keyToken = this.expect(TokenType.Ident);
        this.expect(TokenType.Equals);
        this.skipNewlinesAndComments();
        const value = this.parseValue();
        entries.push({
          type: 'KeyValue',
          key: keyToken.literal,
          value,
          line: keyToken.line,
          col: keyToken.col,
        });
        this.skipComma();
      } else {
        throw new ParseError(
          `Unexpected token '${token.literal}' at ${token.line}:${token.col}`,
          token.line,
          token.col
        );
      }

      this.skipNewlinesAndComments();
    }

    this.expect(TokenType.RBrace);

    return {
      type: 'InlineObject',
      entries,
      line: entries[0]?.line ?? 0,
      col: entries[0]?.col ?? 0,
    };
  }

  private parseValue(): GomlValue {
    this.skipNewlinesAndComments();
    const token = this.current();

    switch (token.type) {
      case TokenType.String:
        this.advance();
        return {
          type: 'String',
          value: token.literal,
          line: token.line,
          col: token.col,
        };

      case TokenType.Number:
        this.advance();
        return {
          type: 'Number',
          value: token.literal.includes('.') ? parseFloat(token.literal) : parseInt(token.literal, 10),
          raw: token.literal,
          line: token.line,
          col: token.col,
        };

      case TokenType.Boolean:
        this.advance();
        return {
          type: 'Boolean',
          value: token.literal === 'true',
          line: token.line,
          col: token.col,
        };

      case TokenType.Null:
        this.advance();
        return {
          type: 'Null',
          line: token.line,
          col: token.col,
        };

      case TokenType.Dollar:
        return this.parseReference();

      case TokenType.LBracket:
        return this.parseArrayValue();

      case TokenType.Ident:
        this.advance();
        return {
          type: 'String',
          value: token.literal,
          line: token.line,
          col: token.col,
        };

      default:
        throw new ParseError(
          `Unexpected token '${token.literal}' at ${token.line}:${token.col}`,
          token.line,
          token.col
        );
    }
  }

  private parseReference(): ReferenceValue {
    const dollarToken = this.expect(TokenType.Dollar);
    let path = '';

    if (this.check(TokenType.Ident)) {
      path = this.advance().literal;
    }

    while (this.check(TokenType.Dot)) {
      this.advance();
      if (this.check(TokenType.Ident)) {
        path += '.' + this.advance().literal;
      }
    }

    return {
      type: 'Reference',
      path: '__ref__:' + path,
      line: dollarToken.line,
      col: dollarToken.col,
    };
  }

  private parseArrayValue(): ArrayValue {
    const bracketToken = this.expect(TokenType.LBracket);
    this.skipNewlinesAndComments();

    const items: GomlValue[] = [];
    while (!this.check(TokenType.RBracket) && !this.isAtEnd()) {
      if (this.check(TokenType.Newline)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.Comma)) {
        this.advance();
        continue;
      }

      if (this.check(TokenType.RBracket)) break;

      items.push(this.parseValue());

      if (this.check(TokenType.Comma)) {
        this.advance();
      }

      this.skipNewlinesAndComments();
    }

    this.expect(TokenType.RBracket);

    return {
      type: 'ArrayValue',
      items,
      line: bracketToken.line,
      col: bracketToken.col,
    };
  }

  private parseComment(): CommentNode {
    const token = this.expect(TokenType.Comment);
    return {
      type: 'Comment',
      text: token.literal,
      line: token.line,
      col: token.col,
    };
  }

  private skipNewlinesAndComments(entries?: GomlNode[]): void {
    while (!this.isAtEnd()) {
      const token = this.current();
      if (token.type === TokenType.Newline) {
        this.advance();
      } else if (token.type === TokenType.Comment) {
        if (entries) {
          entries.push(this.parseComment());
        } else {
          this.advance();
        }
      } else {
        break;
      }
    }
  }

  private skipComma(): void {
    if (this.check(TokenType.Comma)) {
      this.advance();
    }
  }

  private current(): Token {
    if (this.pos >= this.tokens.length) {
      return { type: TokenType.Eof, literal: '', line: 0, col: 0 };
    }
    return this.tokens[this.pos];
  }

  private check(type: TokenType): boolean {
    return this.current().type === type;
  }

  private advance(): Token {
    const token = this.current();
    if (this.pos < this.tokens.length) {
      this.pos++;
    }
    return token;
  }

  private expect(type: TokenType): Token {
    const token = this.current();
    if (token.type !== type) {
      throw new ParseError(
        `Expected ${type} but got ${token.type} '${token.literal}' at ${token.line}:${token.col}`,
        token.line,
        token.col
      );
    }
    return this.advance();
  }

  private isAtEnd(): boolean {
    return this.pos >= this.tokens.length || this.current().type === TokenType.Eof;
  }
}

export function parse(input: string): { [key: string]: any } {
  const parser = new Parser();
  const doc = parser.parse(input);
  return convertDocument(doc);
}

function convertDocument(doc: Document): { [key: string]: any } {
  const result: { [key: string]: any } = {};
  for (const entry of doc.entries) {
    const converted = convertNode(entry);
    if (converted) {
      Object.assign(result, converted);
    }
  }
  return result;
}

function convertNode(node: GomlNode): { [key: string]: any } | null {
  switch (node.type) {
    case 'KeyValue': {
      return { [node.key]: convertValue(node.value) };
    }
    case 'Object': {
      const obj: { [key: string]: any } = {};
      for (const entry of node.entries) {
        const converted = convertNode(entry);
        if (converted) {
          Object.assign(obj, converted);
        }
      }
      return { [node.name]: obj };
    }
    case 'Array': {
      const arr: any[] = node.items.map((item) => convertValue(item));
      return { [node.name]: arr };
    }
    case 'Comment':
      return null;
    default:
      return null;
  }
}

function convertValue(value: GomlValue): any {
  switch (value.type) {
    case 'String':
      return value.value;
    case 'Number':
      return value.value;
    case 'Boolean':
      return value.value;
    case 'Null':
      return null;
    case 'Reference':
      return value.path;
    case 'ArrayValue':
      return value.items.map((item) => convertValue(item));
    case 'InlineObject': {
      const obj: { [key: string]: any } = {};
      for (const entry of value.entries) {
        obj[entry.key] = convertValue(entry.value);
      }
      return obj;
    }
    default:
      return null;
  }
}
